Wrong typescript declarations for `.update()` method (continued)

### What happened
This is a continuation of [issue 1466](https://github.com/immutable-js/immutable-js/issues/1466).
In [1842](https://github.com/immutable-js/immutable-js/pull/1842), `| undefined` was added to the updater function's parameter type, but not the return type. As mentioned in a [comment](https://github.com/immutable-js/immutable-js/issues/1466#issuecomment-353705724) on that issue, the return type also needs to allow `undefined`. This is specifically needed for cases where no update is applied if the entry does not exist, or where the entry should be deleted.

e.g.
```
const myMap = Map<string, List<string>>();
const stillMyMap = myMap.update("noKey", ls => ls?.map(x => x + "blah"));
```
Since `myMap` doesn't have an entry for "noKey", `ls` is `undefined`, and thus the updater returns `undefined`.

### How to reproduce

Run Typescript type-checking on the above example, with `strictNullChecks` turned on. You will get an error like:
```
error TS2345: Argument of type '(ls: List<string>|undefined) => List<string>|undefined' is not assignable to parameter of type  '(ls: List<string>|undefined) => List<string>'.
  Type 'List<string>|undefined' is not assignable to type 'List<string>'.
    Type 'undefined' is not assignable to type 'List<string>'.
```

--- a/type-definitions/immutable.d.ts
+++ b/type-definitions/immutable.d.ts
@@ -424,7 +424,10 @@ declare namespace Immutable {
      * @see `Map#update`
      */
     update(index: number, notSetValue: T, updater: (value: T) => T): this;
-    update(index: number, updater: (value: T | undefined) => T): this;
+    update(
+      index: number,
+      updater: (value: T | undefined) => T | undefined
+    ): this;
     update<R>(updater: (value: this) => R): R;
 
     /**
@@ -1001,7 +1004,7 @@ declare namespace Immutable {
      * Note: `update(key)` can be used in `withMutations`.
      */
     update(key: K, notSetValue: V, updater: (value: V) => V): this;
-    update(key: K, updater: (value: V | undefined) => V): this;
+    update(key: K, updater: (value: V | undefined) => V | undefined): this;
     update<R>(updater: (value: this) => R): R;
 
     /**
@@ -5571,7 +5574,7 @@ declare namespace Immutable {
   function update<K, V, C extends Collection<K, V>>(
     collection: C,
     key: K,
-    updater: (value: V | undefined) => V
+    updater: (value: V | undefined) => V | undefined
   ): C;
   function update<K, V, C extends Collection<K, V>, NSV>(
     collection: C,
@@ -5598,7 +5601,7 @@ declare namespace Immutable {
   function update<V>(
     collection: Array<V>,
     key: number,
-    updater: (value: V) => V
+    updater: (value: V | undefined) => V | undefined
   ): Array<V>;
   function update<V, NSV>(
     collection: Array<V>,

--- a/type-definitions/ts-tests/list.ts
+++ b/type-definitions/ts-tests/list.ts
@@ -265,6 +265,9 @@ import {
   // $ExpectError
   List<number>().update(1, 10, (v: number | undefined) => v + 'a');
 
+  // $ExpectType List<string>
+  List<string>().update(1, (v) => v?.toUpperCase());
+
   // $ExpectType List<number>
   update(List<number>(), 0, (v: number | undefined) => 0);
 

--- a/type-definitions/ts-tests/map.ts
+++ b/type-definitions/ts-tests/map.ts
@@ -176,6 +176,9 @@ import { Map, List } from 'immutable';
 
   // $ExpectError
   Map<number, number>().update(1, 10, (v: number | undefined) => v + 'a');
+
+  // $ExpectType Map<string, string>
+  Map<string, string>().update("noKey", ls => ls?.toUpperCase());
 }
 
 {

